test/e2e/graceful_restart_test.go

```diff
@@ -0,0 +1,57 @@
+package e2e
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/argoproj/argo-cd/v2/pkg/apiclient/settings"
+	"github.com/argoproj/argo-cd/v2/test/e2e/fixture"
+	. "github.com/argoproj/argo-cd/v2/test/e2e/fixture"
+)
+
+func checkHealth(t *testing.T, requireHealthy bool) {
+	t.Helper()
+	resp, err := DoHttpRequest("GET", "/healthz?full=true", "")
+	if requireHealthy {
+		require.NoError(t, err)
+		require.Equal(t, http.StatusOK, resp.StatusCode)
+	} else {
+		if err != nil {
+			if !strings.Contains(err.Error(), "connection refused") && !strings.Contains(err.Error(), "connection reset by peer") {
+				require.NoErrorf(t, err, "If an error returned, it must be about connection refused or reset by peer")
+			}
+		} else {
+			require.Contains(t, []int{http.StatusOK, http.StatusServiceUnavailable}, resp.StatusCode)
+		}
+	}
+}
+
+func TestAPIServerGracefulRestart(t *testing.T) {
+	EnsureCleanState(t)
+
+	// Should be healthy.
+	checkHealth(t, true)
+	// Should trigger API server restart.
+	fixture.SetParamInSettingConfigMap("additionalUrls", "- http://test")
```

Comment:
Modifying additional urls is the safest way to trigger reload that I could find. Feel free to suggest the alternative. I was hesitant to change URL, being afraid it can mess up the things completely. I also wanted to test that server is indeed restarted with new parameters in place (otherwise the test may succeed even if restart didn't happen).

Reply:
That's what it restarts on https://github.com/argoproj/argo-cd/blob/02d6866f8a8276c455986672f2e8709232292e76/server/server.go#L674-L734:
* URL
* AdditionalURLs
* Various secrets/certificates
* Dex/Oidc config
* Extensions
Out of these, additional urls seems to be the only one without potential side effects. As of others, I'm not even sure what effects they can have, but modifying them to some test values doesn't seem promising.